src/source/Crypto/Dtls_mbedtls.c

```diff
@@ -0,0 +1,739 @@
+    mbedtls_ssl_conf_export_keys_ext_cb(&pDtlsSession->sslCtxConfig, dtlsSessionKeyDerivationCallback, pDtlsSession);
+
+    CHK(mbedtls_ssl_setup(&pDtlsSession->sslCtx, &pDtlsSession->sslCtxConfig) == 0, STATUS_SSL_CTX_CREATION_FAILED);
+    mbedtls_ssl_set_mtu(&pDtlsSession->sslCtx, DEFAULT_MTU_SIZE);
+    mbedtls_ssl_set_bio(&pDtlsSession->sslCtx, pDtlsSession, dtlsSessionSendCallback, dtlsSessionReceiveCallback, NULL);
+    mbedtls_ssl_set_timer_cb(&pDtlsSession->sslCtx, &pDtlsSession->transmissionTimer, dtlsSessionSetTimerCallback, dtlsSessionGetTimerCallback);
+
+    // Start non-blocking handshaking
+    pDtlsSession->dtlsSessionStartTime = GETTIME();
+    CHK_STATUS(timerQueueAddTimer(pDtlsSession->timerQueueHandle, DTLS_SESSION_TIMER_START_DELAY,
+                              DTLS_TRANSMISSION_INTERVAL, dtlsTransmissionTimerCallback, (UINT64) pDtlsSession,
+                              &pDtlsSession->timerId));
+
+CleanUp:
+    if (locked) {
+        MUTEX_UNLOCK(pDtlsSession->sslLock);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS dtlsSessionIsInitFinished(PDtlsSession pDtlsSession, PBOOL pIsFinished)
+{
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    CHK(pDtlsSession != NULL && pIsFinished != NULL, STATUS_NULL_ARG);
+    MUTEX_LOCK(pDtlsSession->sslLock);
+    *pIsFinished = pDtlsSession->state == CONNECTED;
+    MUTEX_UNLOCK(pDtlsSession->sslLock);
+
+CleanUp:
+    LEAVES();
+    return retStatus;
+}
+
+STATUS dtlsSessionProcessPacket(PDtlsSession pDtlsSession, PBYTE pData, PINT32 pDataLen)
+{ 
+    ENTERS();
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+    INT32 sslRet, readBytes = 0;
+    PIOBuffer pReadBuffer;
+    CHK(pDtlsSession != NULL && pData != NULL && pData != NULL, STATUS_NULL_ARG);
+    CHK(ATOMIC_LOAD_BOOL(&pDtlsSession->isStarted), STATUS_SSL_PACKET_BEFORE_DTLS_READY);
+    CHK(!ATOMIC_LOAD_BOOL(&pDtlsSession->shutdown), retStatus);
+
+    MUTEX_LOCK(pDtlsSession->sslLock);
+    locked = TRUE;
+
+    pReadBuffer = pDtlsSession->pReadBuffer;
+    CHK_STATUS(ioBufferWrite(pReadBuffer, pData, *pDataLen));
+
+    // read application data
+    while (pReadBuffer->off < pReadBuffer->len) {
+        sslRet = mbedtls_ssl_read(&pDtlsSession->sslCtx, pData + readBytes, pReadBuffer->len - pReadBuffer->off);
+        if (sslRet > 0) {
+            readBytes += sslRet;
+        } else if (sslRet == 0 || sslRet == MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY) {
+            // if sslRet is 0, the connection is closed already.
+            // if sslRet is MBEDTLS_ERR_SSL_PEER_CLOSE_NOTIFY, the client notified us that the connection is going to be closed.
+            // In either case, we'll make sure that the state will change to CLOSED. If it's already closed, it'll be just a noop.
+            DLOGD("Detected DTLS close_notify alert");
+            CHK_STATUS(dtlsSessionShutdown(pDtlsSession));
+            break;
```

Comment:
Best to set a boolean variable that gets checked in the loops condition instead of flow modification. Something like iterate = FALSE, then in the while loop have while (iterate && blah)

Reply:
Done